Update snapshot enhancement for Kubernetes 1.17

jsafrane opened the pull request:
In order to support volume snapshots, OCP needs to run a new cluster-wide snapshot-controller. It manages lifecycle of volume snapshots and is very similar to in-tree PV controller.

OCP needs a new operator to install the snapshot-controller and monitor its health. No new API is required on OCP side, the snapshot-controller will be the same on all clusters and does not require any configuration.

@openshift/storage 
@openshift/openshift-architects

bertinatto commented:
LGTM

bertinatto commented:
/lgtm

deads2k commented on enhancements/storage/csi-snapshot.md:
```diff
@@ -37,52 +38,98 @@ We want to include the upstream CSI Snapshot feature in OpenShift.
+  * Everything runs in namespace `openshift-csi-snapshot-controller`.
+  * Requires new github repo, openshift/csi-snapshot-controller-operator.
+  * Requires new image.
+  * No new CRD / CR is needed for the operator itself, the operator does the same on all clusters / clouds and does not need any configuration.
```
We've found that when building up clusteroperator status it can be useful to have individual FooDegraded and FooAvailable conditions in a myoperator.operator.openshift.io resource and aggregate those conditions for the clusteroperator status.  Persisting the individual conditions makes it very easy to add fully independent loops and allows detailed individual methods.  We use this technique in group-b operators to good effect and have library-go control loops to make the individual wiring easier.

jsafrane replied:
It is worth in this operator that starts just installs CRDs and in the end runs a simple Deployment?

deads2k replied:
> It is worth in this operator that starts just installs CRDs and in the end runs a simple Deployment?

I think so.  It turns out that most operators are impacted by running conditions in the rest of the cluster that are not directly related and the ability to add sanity checks for various degraded condition produces a lot of value by focusing how bugs are redirected and how quickly.  You can imagine sanity checks being built for all sorts of conditions including potentially health of the plugged in units or alerts on too many failures in status for your CRs.   I think the flexibility adds great value for little cost.

group-b even has history adding conditions to operators in other groups.

deads2k commented on enhancements/storage/csi-snapshot.md:
```diff
@@ -37,52 +38,98 @@ We want to include the upstream CSI Snapshot feature in OpenShift.
+  * No new CRD / CR is needed for the operator itself, the operator does the same on all clusters / clouds and does not need any configuration.
 
-We will be releasing the `external-snapshotter` sidecar image to enable CSI Snapshot support in OCP as Developer Preview only. This sidecar can be used by internal teams, such as CNV or OCS, for their drivers.
+2. OCP ships new image csi-snapshot-controller. Its source code is already available in github.com/openshift/csi-external-snapshotter, we only need an image.
```
all of other repos that we fork into openshift start with the original org name first, so we have openshift/kubernetes-apimachinery and openshift/opencontainers-runc.  I'd recommend the same.

jsafrane replied:
Too late, everything we forked from gihub.com/kubernetes-csi has only csi- as prefix, both in repo  and image names. I'd like to keep this tradition to keep things consistent.

deads2k replied:
> Too late, everything we forked from gihub.com/kubernetes-csi has only csi- as prefix, both in repo and image names. I'd like to keep this tradition to keep things consistent.

I don't see it as too late.  We have over a 100 forks.  Fixing the pattern to be consistent with the rest makes sense.  I won't push to rename everything, although automatic redirects will keep them all working if we did.

jsafrane replied:
I've filed https://jira.coreos.com/browse/ART-1366 to rename the repos. Since the repos already exist, it's orthogonal to this PR.

deads2k commented on enhancements/storage/csi-snapshot.md:
```diff
@@ -37,52 +38,98 @@ We want to include the upstream CSI Snapshot feature in OpenShift.
+2. OCP ships new image csi-snapshot-controller. Its source code is already available in github.com/openshift/csi-external-snapshotter, we only need an image.
+  * The component / image is called kubernetes-csi/snapshot-controller upstream, we add csi- prefix to repository and image names, who knows what other kinds of snapshotters there will be.
+
+3. OCP creates upstream snapshot CRDs via CVO.
```
if you think of the CVO as an operator of operators, then its responsibility ends with the creation of your operator's CRDs and workload resources.  The CVO shouldn't be used to lay down the resources of your operator's operand.  Such relationships can be hard to transition and your end up heavily constrained by the CVO. See the issues we had with the CVO stripping our required CRD fields earlier this year.

There are existent helpers for "create all these manifests" in library-go and hopefully in controller-runtime.  I recommend using those.

gnufied replied:
This basically means - we shouldn't be packing CRDs used by snapshot-controller inside snapshot-operator's manifests and rather create them when snapshot-operator is started by using libary-go?

This would also mean - when CRDs change, snapshot-operator should be responsible for upgrading them. I think we need to capture this bit.

jsafrane replied:
> The CVO shouldn't be used to lay down the resources of your operator's operand.

Ack.

deads2k commented on enhancements/storage/csi-snapshot.md:
```diff
@@ -37,52 +38,98 @@ We want to include the upstream CSI Snapshot feature in OpenShift.
+3. OCP creates upstream snapshot CRDs via CVO.
+  * Created during (or before) csi-snapshot-controller-operator installation.
+  * We just copy [upstream CRDs](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/volumesnapshots/crd).
 
```
for debugging, which related resources will you add to your clusteroperator to ensure that you can debug issues?

jsafrane replied:
This is a super simple operator, it just runs one Deployment (+ CRD + maybe ServiceAccount + some RBAC). So gathering the namespace + deployments should do.

Added to the enhancement.

gnufied commented on enhancements/storage/csi-snapshot.md:
```diff
@@ -37,52 +38,98 @@ We want to include the upstream CSI Snapshot feature in OpenShift.
+    * We do not know in advance what CSI drivers will a cluster admin install and we want the cluster ready for snapshots.
+  * This operator will create Deployment with csi-snapshot-controller and report its status via standard `ClusterOperator` object.
+    * Using Deployment with leader election instead of upstream StatefulSet - leader election is significantly faster to run a new leader when a node with the current leader gets unavailable.
+    * TBD: Use DaemonSet on masters? The controller can read / write any VolumeSnapshot, VolumeSnapshotContent and PV objects in the cluster and read any PVC. Running it on masters may be more secure.
```
Daemonset or as static pods?

jsafrane replied:
I talked to @mfojtik, Deployment seems to be the safest option and we can put it on master using node selector + tolerations if we decide so.

jsafrane commented:
Updated the PR

- CRDs will be created by the operator
- Using Deployment with 3 replicas + leader election, for now on regular workers, unless someone tells us otherwise. It's very unlikely that all snapshot controller replicas will be on unavailable nodes - it takes Deployment up to 5 minutes to detect that and create new replicas.
  - Switching to masters is just matter of updating one node selector.
- Added ClusterOperator.Status.RelatedObjects

gnufied commented:
/lgtm

deads2k commented on enhancements/storage/csi-snapshot.md:
```diff
@@ -37,52 +38,98 @@ We want to include the upstream CSI Snapshot feature in OpenShift.
+    * Vanilla copy of [upstream CRDs](https://github.com/kubernetes/kubernetes/tree/master/cluster/addons/volumesnapshots/crd) is used.
+  * This operator will create Deployment with csi-snapshot-controller
+    * Using Deployment with 3 replicas + leader election instead of upstream StatefulSet - leader election is significantly faster to run a new leader when a node with the current leader gets unavailable.
+    * Optionally, run the Deployment on masters using proper node selector + tolerations.
```
This doesn't need to be in a first cut.  The idea that masters are generally special doesn't hold up without strong assumptions on topology that we abandoned in 4.x generally.

deads2k commented:
/approve
